### PR TITLE
Fix several watershed issues (off-center markers, watershed_line)

### LINKED ISSUE #4240: Different label boundaries with skimage.morphology.watershed when using `watershed_line=True` and `watershed_line=False`
## Description
When doing watersheding I get a different boundary between adjacent labels, when using `watershed_line=True` and `watershed_line=False` (see image).
If you compare the two results in the image below, you see that with `watershed_line=True` the watershed lines/label boundaries are somewhat diagonal, whereas with `watershed_line=False` the boundaries are perfectly horizontal.
Is this expected behavior?

Original image:
image

Result from code below:
image

## Way to reproduce
```python
import matplotlib.pyplot as plt
import numpy as np
from skimage.external import tifffile as tff
from skimage.filters import threshold_local, gaussian
from skimage.measure import label
from skimage.morphology import watershed

# read image
imdata = tff.imread("image.png")
image_gaussian_filter_sigma_2 = gaussian(imdata, sigma=2)

# define peaks for watersheding
image_peaks = np.zeros_like(imdata)
pos = (np.array([77, 95, 131, 149, 183, 207, 240, 267, 302, 330]), np.array([51, 50, 50, 50, 50, 50, 50, 50, 50, 49]))
image_peaks[pos] = 1
peaks_labelled = np.flipud(label(np.flipud(image_peaks)))

# create mask for watersheding
adaptive_threshold_vals = threshold_local(image_gaussian_filter_sigma_2, block_size=15)
adative_mask = image_gaussian_filter_sigma_2 > adaptive_threshold_vals

# do watersheding with and without watershed_line
labels_with_watershed_line = watershed(-image_gaussian_filter_sigma_2, markers=peaks_labelled, mask=adative_mask,
                                       watershed_line=True)
labels_without_watershed_line = watershed(-image_gaussian_filter_sigma_2, markers=peaks_labelled, mask=adative_mask,
                                          watershed_line=False)

# display
fig, ax = plt.subplots(1, 3, figsize=[5, 5])
ax[0].imshow(imdata)
ax[1].imshow(labels_with_watershed_line)
ax[2].imshow(labels_without_watershed_line)
plt.show()
```

## Version information
```python
3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
Linux-5.0.0-31-generic-x86_64-with-debian-buster-sid
scikit-image version: 0.15.0
numpy version: 1.17.2
```

### LINKED ISSUE #6632: Off-centre markers for watershed
### Description:

The watershed gives sub-optimal results when the marker is off-centre relative to the maximum of the distance map. I am unsure as to whether this is a bug or to be expected.

If a marker is close to (but not on top of) the maximum of the distance map, I would expect the water to flow towards the max of the distance map, and then continue fill up as normal. However, this is not the case. I have taken the code from the watershed example and offset the markers by 6 pixels in the y-direction. The resulting segmentations are incorrect as shown in the attached image.

### Way to reproduce:

```Python
import numpy as np
import matplotlib.pyplot as plt
from scipy import ndimage as ndi

from skimage.segmentation import watershed
from skimage.feature import peak_local_max

# Generate an initial image with two overlapping circles
x, y = np.indices((80, 80))
x1, y1, x2, y2 = 28, 28, 44, 52
r1, r2 = 16, 20
mask_circle1 = (x - x1)**2 + (y - y1)**2 < r1**2
mask_circle2 = (x - x2)**2 + (y - y2)**2 < r2**2
image = np.logical_or(mask_circle1, mask_circle2)

# Now we want to separate the two objects in image
# Generate the markers as local maxima of the distance to the background
distance = ndi.distance_transform_edt(image)

fig, axes = plt.subplots(ncols=4, figsize=(9, 3), sharex=True, sharey=True, facecolor="white")
ax = axes.ravel()

ax[0].imshow(image, cmap=plt.cm.gray)
ax[0].set_title('Overlapping objects')
ax[1].imshow(-distance, cmap=plt.cm.gray)
ax[1].set_title('Distances')

# shift by 0 or 6 pixels to demonstrate what happens when the markers
# don't perfectly align with the distance map
for i, dy in enumerate((0, 6)):
    coords = peak_local_max(distance, footprint=np.ones((3, 3)), labels=image)
    coords[:, 0] += dy
    mask = np.zeros(distance.shape, dtype=bool)
    mask[tuple(coords.T)] = True
    markers, _ = ndi.label(mask)
    labels = watershed(-distance, markers, mask=image)
    ax[i+2].imshow(labels)
    ax[i+2].imshow(markers, alpha=(markers>0).astype(float))
    ax[i+2].set_title(f'watershed. dy={dy}')

for a in ax:
    a.set_axis_off()

fig.tight_layout()
plt.show()
```

### Traceback or output:

image

### Version information:

```python
>>> import sys; print(sys.version)
3.9.7 (default, Sep 16 2021, 08:50:36) 
[Clang 10.0.0 ]
>>> import platform; print(platform.platform())
macOS-10.16-x86_64-i386-64bit
>>> import skimage; print(f'scikit-image version: {skimage.__version__}')
scikit-image version: 0.19.3
>>> import numpy; print(f'numpy version: {numpy.__version__}')
numpy version: 1.21.2
```

### PR DESCRIPTION
**Pull request by JoOkuma**
## Description

Closes #6632 and closes #4240.
We could try fixing #6279 too, but it will need much more changes, including the heap.
It's not worth it, IMO, because their issue happens in a topology with tie-zones where the watershed solution is not unique and depends on **implementation details**. In the non-tie-zone case, it should be working (#4240).
I suggest closing it with won't fix.

One test had a similar problem to the abovementioned issues, so I did not implement any regression test and only updated the test. It previously expected a line from the watershed like #6632.

By the way, I could not find the minimum double value in cython. This could improve my PR.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
Fix the behavior of `skimage.segmentation.watershed` when
the markers don't align with local minima by making sure
every marker is evaluated before successive pixels.
```
```release-note
Fix edge case where setting `watershed_lines=True`
in `skimage.segmentation.watershed` resulted in an
incorrect solution.
```

### CODE REVIEW

**Comment by jni** [comment]
@JoOkuma all valid Python is valid Cython, so if you can get the min double in Python you can get it in Cython? This is why you often see `import numpy as np; cimport numpy as cnp` in Cython code, to access both.

File: skimage/segmentation/_watershed_cy.pyx
Comment (jni) [inline skimage/segmentation/_watershed_cy.pyx:133]: This value is extremely close to 0 so it is unlikely to be the smallest float64 value. 😂 

```
In [1]: np.finfo(np.float64)
Out[1]: finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)
```

I suggest finding the value outside of the nogil block, `min_float64 = np.finfo(np.float64).min` and using `min_float64` here

**Comment by jni** [comment]
@JoOkuma nice stuff, thank you!!! 🙏🙏🙏🙏

There's just a couple of small things to fix in the comments but otherwise seems ready! Very nice! 😊

**Review by jni — APPROVED** [review]

**Review by lagru — COMMENTED** [review]
Thank you @JoOkuma for tackling this! I am probably missing something but right now I don't understand how this addresses the issues you mentioned. Reproducing examples from both still show the old or similar behavior (see my comment). 

Also I am having a hard time, judging whether the current behavior is actually something we should change and not "just" how the algorithm is supposed to work. Did you make these changes based on a cited source in the docstring or is the source of the changes which results seem more intuitive? In the latter case I am not sure that we do our users a favor by implementing our own custom watershed.

I hope this doesn't seem to harsh a review! I just want to make sure that we don't change things blindly and deviate from the cited sources. The answer to https://github.com/scikit-image/scikit-image/issues/6632 and https://github.com/scikit-image/scikit-image/issues/4240 might very well be just a clearer documentation of the behavior.

File: skimage/segmentation/_watershed_cy.pyx
Comment (JoOkuma) [reply skimage/segmentation/_watershed_cy.pyx]: @lagru, thanks for the review. There was indeed a bug.
#6632 and #4240 were working on commit `6ef145b`, when I fixed the bug pointed out by @jni, they stopped working because of an error.

When I fixed the error, only #6632 worked, which confused me because #4240 was supposed to be fixed too. So I noticed the `_diff_neighbors` was checking the neighbors' labels but not itself, so I fixed it and got #4240 and #6279 working too :)

> You set all markers to the minimal possible double value so that when the actual watershed algorithm starts drawing from the heap, new pixels will always be sorted after the initial markers, regardless of their pixel value?

Yes.

> In that case, I am really curious about the intended effect here. At a first glance it seems un-intuitive to me why markers are a special case whose neighborhood must be evaluated first. 

The watersheding of a marker $p$ will only label pixels with values greater than their own, if you don't set the $p$'s value to $-\infty$ it won't label a neighboring pixel $q$ with lower-value when there's a competing marker $t$ with a path to $q$ lower than $p$'s value. The intended behavior is that the water from the markers should drip down the bottom of neighboring basins (neighboring pixels with lower values).

For reference, [[1]](https://www.researchgate.net/profile/Serge-Beucher/publication/230837870_The_Morphological_Approach_to_Segmentation_The_Watershed_Transformation/links/00b7d5319b26f3ffa2000000/The-Morphological-Approach-to-Segmentation-The-Watershed-Transformation.pdf), Section IV A 4, $$W_{-1}(g) = M$$ The markers are set to level -1 (they consider the image has non-negative values).
This is shown in Figure 16, where the `modified function` (hatched region) is zero on the markers (selected minima).

I changed the compact watershed test again and added a "barrier" with a greater value than the markers' regions, so the result is not an implementation detail as before (tie-zones). I explain it in the test documentation.

Comment (lagru) [reply skimage/segmentation/_watershed_cy.pyx]: Thanks for the detailed response. I haven't finished reviewing everything yet but I can confirm that https://github.com/scikit-image/scikit-image/issues/6632 and https://github.com/scikit-image/scikit-image/issues/4240#issuecomment-1235358536 return the same result regardless of the marker location or the parameter `watershed_line`. Though the output is changed.

Would you be willing to add new tests for both cases to avoid this kind of bug in the future? :pray: E.g. for the `watershed_line` a test that asserts that the labels returned with `watershed_line=True` are a strict subset of the labels returned with `watershed_line=False`? I am happy to do so as well if you are so busy right now!

Comment (JoOkuma) [reply skimage/segmentation/_watershed_cy.pyx]: Hi @lagru, I added the tests. I decreased the offset of the markers from 6 to 4 for the #6632 test.
With 6, it had a 3 pixels difference, it's not much, but I wanted to avoid a magic number in the tests.

It's a tricky example because the difference between the radius and the distance between their center changes the boundary of the results.

My understanding is the following, for two adjacent but not overlapping circles, the markers can have an offset of, at most, the difference between their radii. If it's greater than the radii difference, the neighbors around the marker of the larger circle won't let it propagate before the smaller circle. Note that, this problem is specific to when using EDT for the watershed, and when there's overlap it gets much trickier because it depends on how the EDT changes with the overlap.

File: skimage/segmentation/_watershed_cy.pyx
Comment (lagru) [inline skimage/segmentation/_watershed_cy.pyx:134]: I am still confused about why we need to have this test for `from_minima`. Why not just set
```suggestion
            elem.value = neg_inf
```
always. If markers are identical with local minima, the marker should already be smaller than its neighbors.

I tried this assumption locally but https://github.com/scikit-image/scikit-image/blob/89fa48aabdd96d7b6ae47dd111393564b7488384/skimage/segmentation/tests/test_watershed.py#L515

fails with `assert 1300 == 1301`. I am not sure yet why. :thinking:

Comment (lagru) [reply skimage/segmentation/_watershed_cy.pyx:134]: Hmm, I think this may be because with `elem.value = -np.inf` the push and pop order of the markers is identical, while with `elem.value = image[index]` id depends on the markers value. 

My gut feeling is telling me that we don't want this, because now the output of `watershed(..., markers, ...)` also depends entirely on the given order of the `markers`. Additionally, if the user provides the minima themselves, they may get a different result compared to passing `markers=None` even though the docstring tells them that the result should be the same!

Going from the source cited in https://github.com/scikit-image/scikit-image/pull/7071#discussion_r1271382091, I tend to prefer the following approach: set `elem.value = neg_inf` for all markers and sort the initial marker order by their image value. If the performance penalty due to sorting should be avoided, I am fine with just using the order of given markers, but that should be clearly documented in the docstring. What do you think?

Comment (lagru) [reply skimage/segmentation/_watershed_cy.pyx:134]: Oh, and I tested this, the [failing test](https://github.com/scikit-image/scikit-image/blob/89fa48aabdd96d7b6ae47dd111393564b7488384/skimage/segmentation/tests/test_watershed.py#L515) also fails if the markers are sorted by their value. I assume that the difference is due to similar behavior as in https://github.com/scikit-image/scikit-image/issues/6632...

File: skimage/segmentation/tests/test_watershed.py
Comment (lagru) [inline skimage/segmentation/tests/test_watershed.py:447]: In reply to https://github.com/scikit-image/scikit-image/pull/7071#discussion_r1277815842, thanks for adding the test! Though, I am unsure if I am happy with this I think this actually doesn't cover the case in #6632 qualitatively. To me #6632 is precisely about what happens if one marker spills over into another basin which which I don't think this test covers. So rather than checking if 0 and 4 are identical, I'd be happy with a much simpler case that does qualitatively the same as #6632. E.g. what do you think about checking the output of

```python
image = -np.array(
    [1, 2, 2, 2, 2, 2, 3]
)
markers = np.array(
    [1, 0, 0, 0, 0, 0, 2]
)
ski.segmentation.watershed(image, markers=markers, mask=image != 0)
```
on `main` this gives `[1, 2, 2, 2, 2, 2, 2]`, on your branch this gives `[1, 1, 1, 1, 2, 2, 2]`.

Hope I am not coming across as too nitpicky. I am totally happy to add the tests myself, if you rather focus your time on the actual code. :)

Comment (jni) [reply skimage/segmentation/tests/test_watershed.py:447]: @lagru I like your test suggestion but I'd personally like to keep both — the circles are a more complex landscape for the watershed, so maybe it's not too far fetched to imagine an implementation failing one but not both?

Comment (lagru) [reply skimage/segmentation/tests/test_watershed.py:447]: True :+1:, no reason to waste this test. :D

**Comment by lagru** [comment]
@JoOkuma, feel free to ignore pep8speaks. See https://github.com/scikit-image/scikit-image/pull/7148.

**Review by lagru — CHANGES_REQUESTED** [review]
I still would like for the thread https://github.com/scikit-image/scikit-image/pull/7071#discussion_r1286982436 to be addressed before I think this is ready to merge. The current solution returns different results for identical markers which are local minima depending on whether they are passed explicitly to `watershed` or implicitly with `markers=None`
```python
import numpy as np
import skimage as ski
import scipy.ndimage

image = np.array([-3, -2, -2, -2, -2, -2, -4], dtype=float)

# User supplies local minima
minima = ski.morphology.local_minima(image)
markers = scipy.ndimage.label(minima)[0]
result_1 = ski.segmentation.watershed(image, markers=markers)

# Minima are determined internally
result_2 = ski.segmentation.watershed(image)

np.testing.assert_array_equal(result_1, result_2)
# Mismatched elements: 1 / 7 (14.3%)
# x: array([1, 1, 1, 1, 2, 2, 2], dtype=int32)
# y: array([1, 1, 1, 2, 2, 2, 2], dtype=int32)
```

I think that would be a confusing API.

**Comment by JoOkuma** [comment]
Hi @lagru, sorry about being a bit absent. I don't have the bandwidth to track this PR.
I can get back to it in late November. In the meantime, feel free to make any changes you think are necessary.

In the example you showed above, both results are valid watershed results because of tie-zones, it isn't evaluating the correctness of the algorithm.
Nonetheless, I understand that you expect them to have the same results because skimage's watershed from minima uses the same markers, and the suggested change makes sense.

**Comment by mkcor** [comment]
@lagru wrote:

> The current solution returns different results for identical markers which are local minima depending on whether they are passed explicitly to `watershed` or implicitly with `markers=None`

This very issue is already present and 'fixing' it "would certainly be something for skimage2" (extrapolating your https://github.com/scikit-image/scikit-image/issues/7117#issuecomment-1736151352 here) so I would argue it is independent of this PR. In other words, I think this PR shouldn't be judged in terms of this criterion of consistency (and I definitely agree that this inconsistency is an issue).

I just wanted to point to the conversation which happened over at https://github.com/scikit-image/scikit-image/issues/7117, because I felt it was overlooked (even contradicted?) when I read your 'requested changes' review.

**Comment by mkcor** [comment]
> If you run the snippet on current `main` it passes.

@lagru right. My bad. If we replace
```py
image = np.array([-3, -2, -2, -2, -2, -2, -4], dtype=float)
```
with a *real* image, say,
```py
image = ski.data.camera()
```
it fails. Let's see:
```py
import numpy as np
import skimage as ski
import scipy.ndimage

image = ski.data.camera()

# User supplies local minima
minima = ski.morphology.local_minima(image)
markers_1 = scipy.ndimage.label(minima)[0]
print('number of markers (nd.im) :', markers_1.max())
result_1 = ski.segmentation.watershed(image, markers=markers_1)

# User supplies local minima using morphology label function
markers_1b = ski.morphology.label(ski.morphology.local_minima(image))
print('number of markers (morph) :', markers_1b.max())
result_1b = ski.segmentation.watershed(image, markers=markers_1b)

# Internally, default markers are computed as
connectivity = 2
mask = np.ones(image.shape, dtype=bool)
markers_bool = ski.morphology.local_minima(image, connectivity=connectivity) * mask
footprint = scipy.ndimage.generate_binary_structure(markers_bool.ndim, connectivity)
markers_2 = scipy.ndimage.label(markers_bool, structure=footprint)[0]
print('number of markers (def/c) :', markers_2.max())

# Minima are determined internally
result_2 = ski.segmentation.watershed(image, connectivity=2)

#np.testing.assert_array_equal(markers_1, markers_2)  # fails
#np.testing.assert_array_equal(result_1, result_2)  # fails (as an expected consequence)

np.testing.assert_array_equal(markers_1b, markers_2)  # passes
#np.testing.assert_array_equal(result_1b, result_2)  # fails "less" (~20% mismatch vs. 99%)
```
Output being:

> number of markers (nd.im) : 14520
> number of markers (morph) : 13563
> number of markers (def/c) : 13563

So, unlike I once thought (oral communication), it's not enough to pass the same markers (`markers_1b`) as those computed internally with `connectivity=2` (`markers_2`) to obtain the same segmentation result; it gets close(r) but there must be other computation steps involving the various parameters.

**Comment by mkcor** [comment]
pre-commit.ci autofix

**Comment by JoOkuma** [comment]
Hey, @lagru and @mkcor, 

I removed the `from_minima` parameter, as suggested by [this comment](https://github.com/scikit-image/scikit-image/pull/7071#discussion_r1286982436), as it was indeed unnecessary.

I had to adjust the test because of an off-by-one error shown by @lagru. It's related to a tie-zone, and it doesn't indicate that the algorithm is wrong.

image

This is a screenshot of the pixel (in red) that caused the test to fail, with the segmentation labels overlaid over the original watershed topology.

I noticed a bug @mkcor example, `result_2` is run with `connectivity=2` and `result_b1` with `connectivity=1`, when setting `result_1b = ski.segmentation.watershed(image, markers=markers_1b, connectivity=2)` the comparison between `result_2` and `result_1b` passes.

I think all the comments have been addressed, if I missed anything or additional changes are needed please let me know.

**Comment by mkcor** [comment]
Oh, thank you, @JoOkuma! You got my intention right. :pray: 

So, in the end, @lagru, we *do* 'land on our feet' as long as we use a connectivity of 2 in 2D, i.e., a neighborhood of 8 pixels. :relaxed: Re-copying-pasting the corrected code here (using `connectivity=2` everywhere for the comparisons to make sense):
```py 
import numpy as np
import skimage as ski
import scipy.ndimage

image = ski.data.camera()

# Use connectivity of 2 in 2D, i.e., neighborhood of 8 pixels
connectivity = 2

# User supplies local minima
minima = ski.morphology.local_minima(image)
markers_1 = scipy.ndimage.label(minima)[0]
print('number of markers (nd.im) :', markers_1.max())
result_1 = ski.segmentation.watershed(image, markers=markers_1, connectivity=connectivity)

# User supplies local minima using morphology label function
markers_1b = ski.morphology.label(ski.morphology.local_minima(image))
print('number of markers (morph) :', markers_1b.max())
result_1b = ski.segmentation.watershed(image, markers=markers_1b, connectivity=connectivity)

# Internally, default markers are computed as
mask = np.ones(image.shape, dtype=bool)
markers_bool = ski.morphology.local_minima(image, connectivity=connectivity) * mask
footprint = scipy.ndimage.generate_binary_structure(markers_bool.ndim, connectivity)
markers_2 = scipy.ndimage.label(markers_bool, structure=footprint)[0]
print('number of markers (def/c) :', markers_2.max())

# Minima are determined internally
result_2 = ski.segmentation.watershed(image, connectivity=connectivity)

#np.testing.assert_array_equal(markers_1, markers_2)  # fails
#np.testing.assert_array_equal(result_1, result_2)  # fails (as an expected consequence)

np.testing.assert_array_equal(markers_1b, markers_2)  # passes
np.testing.assert_array_equal(result_1b, result_2)  # passes
```

**Comment by jni** [comment]
@lagru, both you and @JoOkuma are at SciPy so it seems like a good time to get this merged. 😃

File: skimage/segmentation/_watershed_cy.pyx
Comment (lagru) [reply skimage/segmentation/_watershed_cy.pyx]: I updated this test in https://github.com/scikit-image/scikit-image/pull/7071/commits/1e3b8ed2a156603b6619b293e5fb2b87521e44c0 to use the actual bug scenario with an y-offset of 6. Instead of comparing to the 0-offset case, I'm checking the result with regionproperty measures eccentricity and num_pixels.

**Review by lagru — APPROVED** [review]
This finally ready to approve this. :sweat_smile: Thank you @JoOkuma for bearing with us (me).

**Comment by JoOkuma** [comment]
hey @jni, I think @lagru is waiting for your final approval.

File: skimage/segmentation/_watershed_cy.pyx
Comment (jni) [inline skimage/segmentation/_watershed_cy.pyx:152]: I'm confused — `wsl` is not checked at all now?

**Review by jni — COMMENTED** [review]
You guys shouldn't wait for me cos I won't have a chance to do a really thorough review for at least a month. I did a quick read through and had a question which maybe can be clarified in a comment, but otherwise I leave this to you guys to assess. 🙏 

Maybe @lagru can walk @stefanv through the PR at EuroSciPy this coming week for another pair of eyes/sanity check.

Thanks all!

**Comment by stefanv** [comment]
@JoOkuma Thank you very much for this valuable contribution! I think we should go ahead and merge.

May I ask you one more favor? I would love to have a thorough description of the fixes this PR introduces. Could you provide me with a detailed commit message to use? Lars explained the changes to us at the EuroSciPy sprint, and it feels like the kind of subtleties we should clearly document for future reference.

**Comment by stefanv** [comment]
(And maybe answer/resolve @jni's comment, please.)

**Comment by JoOkuma** [comment]
Thanks, @stefanv and @lagru for the input.
I'm in a bit of a crunch in the next weeks, but I'll get back to this soon to finish this PR.

**Comment by mkcor** [comment]
@lagru I was about to take a look at this PR but, in light of https://github.com/scikit-image/scikit-image/pull/7071#issuecomment-2350271161, I'll wait for @JoOkuma's next changes.

**Comment by lagru** [comment]
@mkcor, I think that only pertains to writing the release note(s) in the PR description. Please feel welcome to give it a look. :pray: Even if I misunderstand, I don't expect any sweeping or significant changes to this PR. :wink:

**Comment by lagru** [comment]
Thanks @JoOkuma and everyone else for finally getting this across the line! :partying_face:

**Comment by JoOkuma** [comment]
Hi @lagru, I missed you merged this PR.
Thanks a lot, and happy holidays.

**Pull request merged**
